lib/archethic/contracts/interpreter.ex

```diff
@@ -136,6 +197,179 @@ defmodule Archethic.Contracts.Interpreter do
   #  | .__/|_|  |_| \_/ \__,_|\__\___|
   #  |_|
   # ------------------------------------------------------------
+
+  defp do_execute(
+         :transaction,
+         trigger_code,
+         contract,
+         incoming_tx = %Transaction{},
+         %Contract{version: version, conditions: conditions},
+         opts
+       ) do
+    constants = %{
+      "transaction" => Constants.from_transaction(incoming_tx),
+      "contract" => contract.constants.contract
+    }
+
+    if valid_conditions?(version, conditions.transaction, constants) do
+      execute_trigger_and_validate_inherit_condition(
+        version,
+        trigger_code,
+        contract,
+        incoming_tx,
+        conditions.inherit,
+        opts
+      )
+    else
+      {:error, :invalid_transaction_constraints}
+    end
+  end
+
+  defp do_execute(
+         :oracle,
+         trigger_code,
+         contract,
+         oracle_tx = %Transaction{},
+         %Contract{version: version, conditions: conditions},
+         opts
+       ) do
+    constants = %{
+      "transaction" => Constants.from_transaction(oracle_tx),
+      "contract" => contract.constants.contract
+    }
+
+    if valid_conditions?(version, conditions.oracle, constants) do
+      execute_trigger_and_validate_inherit_condition(
+        version,
+        trigger_code,
+        contract,
+        oracle_tx,
+        conditions.inherit,
+        opts
+      )
+    else
+      {:error, :invalid_oracle_constraints}
+    end
+  end
+
+  defp do_execute(
+         _trigger_type,
+         trigger_code,
+         contract,
+         nil,
+         %Contract{version: version, conditions: conditions},
+         opts
+       ) do
+    execute_trigger_and_validate_inherit_condition(
+      version,
+      trigger_code,
+      contract,
+      nil,
+      conditions.inherit,
+      opts
+    )
+  end
+
+  defp execute_trigger_and_validate_inherit_condition(
```

Comment:
Maybe this function can be split into two: one for trigger execution and one for the inherit acceptance.